Add Documentation (#664)
zio/zio-prelude: docs/functionaldatatypes/zset.md

Conceptually, we can think of a `ZSet[A, B]` as a `Map[A, B]`, where `A` represents the elements of the set and `B` represents some notion of how many times each element appears.

We can then define a `Set` as a `ZSet` where the measure of how many times each element appears is a `Boolean`.

```scala mdoc
import zio.prelude.ZSet

type Set[+A] = ZSet[A, Boolean]
```

That is, in a `Set` the measure of how many times an element appears is just yes or no. Either the element is in the set or it is not.

However, we can also define many other variants of sets by choosing different types to measure how many times an element appears.

For example, by using the `Natural` new type from ZIO Prelude, which describes integers greater than or equal to zero, we can define the type of a `MultiSet` or "bag" in some programming languages.

```scala mdoc:reset
import zio.prelude.ZSet
import zio.prelude.newtypes.Natural

type MultiSet[+A] = ZSet[A, Natural]
```

This ends up being a quite useful data type and can represent anything where the number of times an element appears matters but the order doesn't matter, such as a shopping cart.

> By using other measures of how many times an item appears we can describe other variants of sets such as `Int` for sets where items can appear a negative number of times or `Double` for sets where the number of times an item appears can be fractional.

Reviewer: semi-pedantic: I believe you but this usage is rather novel to me, maybe an external reference to common usage of such idioms (or a quote)? I can imagine a debt of $5.03 being a negative double set of dollars but I am not sure I would want to model that as a "generalized" set.

`ZSet` is focused on using the functional abstractions and new types in ZIO Prelude to describe operators on these generalizations of sets in as practical a way as possible.

## Constructing ZSets

The most basic way to construct a `ZSet` is from a `Map` where the keys in the map represent the elements in the `ZSet` and the values in the map represent the number of times each element appears.

```scala mdoc
val shoppingList: ZSet[String, Int] =
  ZSet.fromMap(Map("apples" -> 2, "oranges" -> 3, "bananas" -> 5))
```

We can also construct a `ZSet` from existing collection types.

If we have a normal `Set` from the Scala standard library, we can convert it to a `ZSet` using the `fromSet` operator.

```scala mdoc
val set: ZSet[Int, Boolean] =
  ZSet.fromSet(Set(1, 2, 3))
```

If we have an `Iterable` we can convert it into a `ZSet` using the `fromIterable` constructor. The measure of how many times each item appears will just be how many times it appears in the `Iterable`.

```scala mdoc
val multiSet: MultiSet[Int] =
 ZSet.fromIterable(List(1, 1, 2))
```

> Finally, we can just construct an empty `ZSet` using the `empty` operator and add elements to it later.

Reviewer: nitpick on "add elements to it later" as we're in a immutable data structure mindset, maybe some rewording could take place?
For example ... "and combine this empty `ZSet` with some non empty `ZSet` later"?

Reviewer: ```suggestion
Finally, we can just construct an empty `ZSet` using the `empty` operator and combine this empty ZSet with some non empty ZSet later.
```
(to avoid implying mutation)

```scala mdoc
val empty: ZSet[Nothing, Nothing] =
  ZSet.empty
```

## Operators on ZSets

As a generalization of sets, `ZSet` supports many of the same operators that exist on sets.

### Set Membership

We can determine how many times an element appears in the set using the `apply` operator. This requires that the measure type has an `Identity` defined for it.

If the element exists in the set the measure of how many times it appears will be returned. Otherwise the identity element will be returned.

> Thus, `shoppingList("apples)` will return `2` and `shoppingList("eggs")` will return `0` while `set(1)` will return `true` and `set(4)` will return `false`.

Reviewer: missing trailing double quote in `("apples)`

Reviewer: ```suggestion
Thus, `shoppingList("apples")` will return `2` and `shoppingList("eggs")` will return `0` while `set(1)` will return `true` and `set(4)` will return `false`.
```

```scala mdoc
val left: ZSet[Int, Int] =
  ZSet.fromMap(Map(1 -> 2, 2 -> 3, 3 -> 4))

val right: ZSet[Int, Int] =
  ZSet.fromMap(Map(2 -> 2, 3 -> 3, 4 -> 4))

val union: ZSet[Int, Int] =
  left.union(right)
```

Here the result will be `ZSet(1 -> 2, 2 -> 3, 3 -> 4, 4 -> 4)` because all elements in both sets will be included in the union of the two sets.

## Set Intersection

Set intersection is another basic way of combining sets.

The intersection of two sets only contains the elements that appear in both sets. If an element appears in both sets the measure of how many times it appears in the intersection will be the minimum of the two measures, as determined by an `Ord` defined on the measure type.

```scala mdoc:nest
val left: ZSet[Int, Int] =
  ZSet.fromMap(Map(1 -> 2, 2 -> 3, 3 -> 4))

val right: ZSet[Int, Int] =
  ZSet.fromMap(Map(2 -> 2, 3 -> 3, 4 -> 4))

val intersection: ZSet[Int, Int] =
  left.intersect(right)
```

The result will be `ZSet(2 -> 2, 3 -> 3)` because the only commonality of the two sets is that they contain these elements.

## Set Difference

We can also compute the difference of two sets.

> The difference of two sets is obtained by subtracting the measure of how many times each element appears in the second set from how many times it appears in the first set. This requires an `Inverse` defined on the `Sum` of the measure type to tell us what it means to "subtract" for this type.

Reviewer: I like this statement as it definitely encourages users to have "algebraic" reasoning/mindset when using Prelude abstractions.

Author: Thanks!

```scala mdoc:nest
val left: ZSet[Int, Int] =
  ZSet.fromMap(Map(1 -> 2, 2 -> 3, 3 -> 4))

val right: ZSet[Int, Int] =
  ZSet.fromMap(Map(2 -> 2, 3 -> 3, 4 -> 4))

val diff: ZSet[Int, Int] =
  left.diff(right)
```

The result will be `ZSet(1 -> 2, 2 -> 1, 3 -> 1, 4 -> -4)`.

Note that the measure of how many times `4` appears in the resulting set is negative. This is permissible because `Int` allows negative numbers.

If we didn't want this we could define the measure type to be a type like `Natural` that did not permit negative numbers.

```scala mdoc:nest
val left: MultiSet[String] =
  ZSet.fromIterable(List("apple", "banana"))

val right: MultiSet[String] =
  ZSet.fromIterable(List("banana", "banana", "banana"))

val diff: MultiSet[String] =
  left.diff(right)
```

> Now the measure of how many times each item appears is a natural number so "banana" cannot appear a negative number of times.

Reviewer: I also like this.

    )
  )
```

The product of this set with itself represents the set of all possible outcomes of rolling two dice.

```scala mdoc
val pair: ZSet[(Int, Int), Double] =
  die.zip(die)
```

This set has thirty six elements, corresponding to each possible combination of outcomes from the two dice rolls, each with a probability of one in thirty six.

This gets more interesting if we say that matters is not the individual dice rolls but their sum.

We can express this in code with the `zipWith` operator, which allows us to combine the elements from the product of the two sets with a function. If multiple elements from the product are mapped to the same value the measures of how many times each value appears will be added, using a `Commutative` instance defined on the `Sum` of the measure type.

```scala mdoc
val sum: ZSet[Int, Double] =
  die.zipWith(die)(_ + _)
```

The resulting set has eleven elements, corresponding to the possible sums from two dice rolls of two to twelve. The measure for each value corresponds to the probability of that sum being rolled.

So for example the probability of seven is about 16.7% whereas the probability of rolling two or twelve is only about 2.8%.

> We can take this even further using the `flatMap`.

Reviewer: ```suggestion
We can take this even further using `flatMap`.
```

Author: Thanks!

The `zipWith` operator lets us combine two sets when the sets are independent of each other. With the `flatMap` operator we can create a new set based on each element of the first set and then "flatten" the resulting set down.

> We could rewrite the example above using `flatMap` and a _for comprehension_ like this:

Reviewer: nitpick: I don't like the style of `for comprehension` here.

Author: Changed.

```scala mdoc:nest
val sum: ZSet[Int, Double] =
  for {
    a <- die
    b <- die
  } yield a + b
```

This allows us to express conditional logic, such as the probability distribution of outcomes for a subsequent event depending on the outcome of a previous event.

## Transforming Elements

We can transform set elements using the `map` operator.

For example, we could add one to each value in a set like this:

```scala mdoc
val setPlusOne: ZSet[Int, Boolean] =
  set.map(_ + 1)
```

This requires a `Commutative` instance defined for the `Sum` of the measure type so that we can combine the measures if we map two elements to the same value.

> ## Transforming The Measure

Reviewer: ```suggestion
## Transforming the Measure
```

Author: Changed.

We can also transform the measure we use to represent how many times an item appears using the `transform` operator. For example, perhaps we have constructed a set where the measure is an `Int` but we would like to transform it to a `Double` so we can compose it with other `ZSet` values with `Double` as the measure type.

We can do that with `transform` like this:

```scala mdoc
def constant(n: Int): ZSet[Int, Double] =
  ZSet(n).transform(_.toDouble)
```

## Extracting ZSet Values

If the number of times a value appears in the set is a natural number then we can use normal collection operators to fold over its values, treating it much like a normal set

```scala mdoc
val fruits: MultiSet[String] =
  ZSet.fromIterable(List("apple", "banana", "orange", "apple", "orange"))

val fruitCount: Int =
  fruits.foldLeft(0)((n, _) => n + 1)
```

This will return `5` because there are five elements in the set we are folding over.

We can convert a `ZSet` back to a `Map` from elements to a measure of how many times they appear using the `toMap` operator.

> We can also always convert a `ZSet` back to a `Set` using the `toSet` operator. Any elements in the `ZSet` with a measure equal to zero based on an `Identity` defined on the `Sum` of the measure type will be discarded from the resulting set.

Reviewer: hopefully this works well for a Double measures and numbers almost equal to 0.0 but distinct from 0.0.

Author: Yeah maybe we should require that the measure type be `Boolean` to convert it back so the user has to provide a function. I do think if the user is using a data type that is subject to floating point precision like that they are probably aware of the issues with it.
